Tracer performance optimisations (solution 1)

ziogaschr opened the pull request:
This resolves #254.

This PR handles tracer performance optimisations and it is one of the two suggested solutions which has:

- cleaner logic in the `tracer.go` file, as it has no custom logic there, based on a JS tracer
- slower performance than solution 2, as it still has to call a JS method using duktape

Below you can find some sample measurements for each test case, with the important one being the `bigSlow`:

### Test CallTracer

| Test | not optimised | optimised |
|------|-------------:|-----------:|
| bigSlow | 5.45653184s | 3.339744961s |
| create | 2.575857ms | 1.721851ms |
| deepCalls | 69.298277ms | 53.81494ms |
| delegatecall | 13.072359ms | 12.216448ms |
| innerCreateOogOuterThrow | 4.356151ms | 5.754842ms |
| innerInstafail | 1.789353ms | 3.336143ms |
| innerThrowOuterRevert | 15.972036ms | 13.377374ms |
| oog | 5.654414ms | 6.051736ms |
| revert | 2.175ms | 1.472202ms |
| revertReason | 4.936105ms | 4.174434ms |
| simple | 5.427691ms | 3.436775ms |
| throw | 1.909941ms | 3.426322ms |

### Test CallTracerParity

| Test | not optimised | optimised |
|------|-------------:|-----------:|
| bigSlow | 5.377376866s | 4.381227684s |
| create | 2.487462ms | 1.834985ms |
| deepCalls | 80.946559ms | 66.649341ms |
| delegatecall | 14.748616ms | 17.8709ms |
| innerCreateOogOuterThrow | 10.448277ms | 9.212009ms |
| innerInstafail | 2.620816ms | 2.585761ms |
| innerThrowOuterRevert | 15.346743ms | 14.264673ms |
| oog | 5.349929ms | 2.512547ms |
| revertReason | 4.295931ms | 728.27µs |
| revert | 1.33ms | 3.595533ms |
| simple | 4.94809ms | 7.79357ms |
| throw | 3.851077ms | 5.416296ms |

ziogaschr committed:
eth/tracers: optimise parity tracer checks for undefined vals for duktape

based on duktape docs checking for `void 0` helps more

ziogaschr committed:
eth/tracers: add test case for `call_tracer` for #254

ziogaschr committed:
eth/tracers: fix makeTest snippet for generating new test cases

ziogaschr committed:
eth/tracer: add test runner for `call_tracer_parity` tests + add tests

ziogaschr committed:
eth/tracers: apply call_tracer* optimisations for bigger transactions

ziogaschr committed:
generate assets.go

ziogaschr committed:
eth/tracers: cleanup op error on `CaptureFault`

ziogaschr committed:
generate assets.go

meowsbits commented on eth/tracers/internal/tracers/call_tracer_parity.js:
```diff
@@ -50,11 +50,30 @@
 		// Capture any errors immediately
 		var error = log.getError();
-		if (typeof error !== "undefined") {
+		if (error !== undefined) {
```
There's a lot of these `undefined` vs. `"undefined"` comparison changes, and they're also made at https://github.com/etclabscore/core-geth/pull/256/files. Is this important to both proposals?  Or could these be extracted to a separate refactoring/optimization change set?

ziogaschr replied:
They are on both proposals.It could be on a separate PR, but I didn't feel the need to separate this single commit. Let me know if you prefer it that way.

meowsbits commented on eth/tracers/tracers_test.go:
```diff
@@ -298,3 +344,103 @@ func jsonEqual(x, y interface{}) bool {
+
+// jsonEqual is similar to reflect.DeepEqual, but does a 'bounce' via json prior to
+// comparison
+func jsonEqualParity(x, y interface{}) bool {
```
What is the reasoning/motivation behind the "bounce" in this function?

ziogaschr replied:
At this point I just copied the existing `jsonEqual` method, later on we might use a single one and reflect to get the wanted type.

The bounce I think is that it takes a `json.RawMessage` and unmarshals it in a go type in order to do `reflect.DeepEqual`

meowsbits commented on eth/tracers/tracers_test.go:
```diff
@@ -298,3 +344,103 @@ func jsonEqual(x, y interface{}) bool {
+		t.Fatalf("failed to retrieve tracer test suite: %v", err)
+	}
+	for _, file := range files {
+		if !strings.HasPrefix(file.Name(), "parity_call_tracer_") {
```
Instead of `ioutil.Readdir` followed by this filter in the next stanza, it might be more legible to use `filepath.Glob("testdata/parity_call_tracer*")`.

ziogaschr replied:
Wonderful idea. At this point I tried to do the minimal changes from existing code in order to later make a PR at foundation/upstream. This test is an almost exact copy of https://github.com/etclabscore/core-geth/pull/255/files/b194b2c4f83ae3f15927accec95e01bc391e68fe#diff-6b1dc15800ea95ec572b25668c1a2f58ebae907d561623718f54e6a6cb7977e7R251

Do you think keeping a copy without additional changes helps on PRs to upstream?

meowsbits replied:
Gotch, yeap, that sounds fine to me. Maybe we can do a tiny `nice-to-have` PR for the globber later ;)

meowsbits commented on eth/tracers/testdata/call_tracer_big_slow.json:
```diff
@@ -0,0 +1,81 @@
+{
```
Gosh, how it warms my heart to see these tests! :pray:

meowsbits replied:
How did you generate these? Worth considering adding the generation program to the source somehow?

ziogaschr replied:
It's very easy. You can see the actual script [here](https://github.com/etclabscore/core-geth/pull/255/files/b194b2c4f83ae3f15927accec95e01bc391e68fe#diff-6b1dc15800ea95ec572b25668c1a2f58ebae907d561623718f54e6a6cb7977e7R47-R93)

meowsbits replied:
Oh, I see... nice.

```
// To generate a new callTracer test, copy paste the makeTest method below into
// a Geth console and call it with a transaction hash you which to export.
```
```
// makeTest generates a callTracer test by running a prestate reassembled and a
// call trace run, assembling all the gathered information into a test case.
```

ziogaschr committed:
eth/tracers: add benchmark for call_tracer_parity

ziogaschr committed:
eth/tracers: add more tests for call_tracer_parity

ziogaschr committed:
eth/tracers: add more tests for call_tracer_parity

ziogaschr committed:
eth/tracers: set go CaptureState before optimisation in order is set for CaptureFault

While this is not an issue when running either the existing tests or external ones, logically it might cause issues if the CaptureFault path is being triggered.

For this reason, we always set the CaptureState context, even by decreasing the performance for a tiny fraction
